File: packages/contracts/src/certification.ts
import type { VersionReference } from "./lifecycle.js";

/** Technical definition conformance only; this does not authorize business use. */
export interface CertificationDefinition {
  claim: "CANONICAL_DEFINITION_CONFORMANCE";
  evaluator: "canonical-structural-contract/v1";
  subject_type: string;
  required_checks: Array<"schema compatibility" | "identity cycles" | "dependency version pins" | "impact">;
  meaning: string;
  limitations: string;
}

export interface CertificationContract {
  definition: CertificationDefinition;
  subject_schema_id?: string | null;
  evidence_id?: string | null;
}

export interface CertificationEvaluationRequest {
  request_id?: string;
  subject: VersionReference;
  contract: VersionReference;
}

export interface DefinitionConformanceReceipt {
  receipt_id: string;
  proof_hash: string;
  recorded_at: string;
  current_use_authorized: false;
  proof: {
    purpose: "CANONICAL_DEFINITION_CONFORMANCE";
    evaluator: "canonical-structural-contract/v1";
    status: "PASS";
    subject: VersionReference;
    contract: VersionReference;
    subject_content_hash: string;
    contract_content_hash: string;
    access_entity: string;
    contract_attributes: CertificationContract;
    subject_schema: VersionReference | null;
    promotion_proposal_id: string;
    promotion_evaluation: {
      evaluator: string;
      proposal_hash: string;
      binding_hash: string;
      status: "PASS";
      expectations: Array<{name: string; status: "PASS" | "FAIL"}>;
      recorded_at: string;
      checks: string[];
      scope: string;
    };
    subject_upstream: ConformanceDependency[];
    contract_upstream: ConformanceDependency[];
  };
}

export interface ConformanceDependency extends VersionReference {
  content_hash: string;
  access_entity: string;
  event_id: string | null;
}

File: services/api/src/finai_api/domain/certification.py
"""Explicit, bounded structural certification; never financial certification."""

from typing import Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field, model_validator

from finai_api.domain.ontology_definitions import DEFINITION_MODELS
from finai_api.domain.resource_lifecycle import VersionReference

BOOTSTRAP_TYPES = frozenset({"SchemaDefinition", "SemanticContract", "LinkType"})
SUBJECT_TYPES = frozenset(DEFINITION_MODELS) | BOOTSTRAP_TYPES
Check = Literal["schema compatibility", "identity cycles", "dependency version pins", "impact"]


class CertificationDefinition(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)
    claim: Literal["CANONICAL_DEFINITION_CONFORMANCE"]
    evaluator: Literal["canonical-structural-contract/v1"]
    subject_type: str
    required_checks: list[Check] = Field(min_length=1, max_length=4)
    meaning: str = Field(min_length=10, max_length=2000)
    limitations: str = Field(min_length=10, max_length=2000)

    @model_validator(mode="after")
    def supported(self) -> "CertificationDefinition":
        if self.subject_type not in SUBJECT_TYPES:
            raise ValueError("Only canonical definition conformance is supported")
        if len(set(self.required_checks)) != len(self.required_checks):
            raise ValueError("Certification checks must be unique")
        return self


class CertificationContract(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)
    definition: CertificationDefinition
    subject_schema_id: UUID | None = None
    evidence_id: UUID | None = None

    @model_validator(mode="after")
    def schema_required(self) -> "CertificationContract":
        if self.subject_schema_id is None and self.definition.subject_type not in BOOTSTRAP_TYPES:
            raise ValueError("Non-bootstrap subjects require an exact canonical schema dependency")
        return self


class CertificationEvaluationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)
    request_id: UUID = Field(default_factory=uuid4)
    subject: VersionReference
    contract: VersionReference
